Index: libdwarf/libdwarf_info.c
===================================================================
--- libdwarf/libdwarf_info.c	(revision 2517)
+++ libdwarf/libdwarf_info.c	(working copy)
@@ -144,7 +144,12 @@
 		/* Add the compilation unit to the list. */
 		STAILQ_INSERT_TAIL(&dbg->dbg_cu, cu, cu_next);
 
-		if (cu->cu_version != 2 && cu->cu_version != 3) {
+		if (cu->cu_version != 2 && cu->cu_version != 3
+                    /* XXX DRi#797: seems to work (not many changes in 4)
+                     * so going to just allow it until fixed upstream via
+                     * https://sourceforge.net/apps/trac/elftoolchain/ticket/394
+                     */
+                    && cu->cu_version != 4) {
 			DWARF_SET_ERROR(dbg, error,
 			    DW_DLE_VERSION_STAMP_ERROR);
 			ret = DW_DLE_VERSION_STAMP_ERROR;
