diff --git a/libelftc/elftc_bfdtarget.c b/libelftc/elftc_bfdtarget.c
index 75e8692d..f56b977c 100644
--- a/libelftc/elftc_bfdtarget.c
+++ b/libelftc/elftc_bfdtarget.c
@@ -26,3 +26,7 @@
 
+#if !(defined _WIN32 && ! defined __CYGWIN__)
 #include <sys/param.h>
+#else
+#include "windows-mingw/sys_param.h"
+#endif
 #include <string.h>
diff --git a/libelftc/elftc_demangle.c b/libelftc/elftc_demangle.c
index d5039d6e..3ffe0305 100644
--- a/libelftc/elftc_demangle.c
+++ b/libelftc/elftc_demangle.c
@@ -26,3 +26,7 @@
 
+#if !(defined _WIN32 && ! defined __CYGWIN__)
 #include <sys/param.h>
+#else
+#include "windows-mingw/sys_param.h"
+#endif
 #include <assert.h>
diff --git a/libelftc/elftc_string_table.c b/libelftc/elftc_string_table.c
index 2f8a6501..095a2a23 100644
--- a/libelftc/elftc_string_table.c
+++ b/libelftc/elftc_string_table.c
@@ -26,4 +26,12 @@
 
+#if !(defined _WIN32 && ! defined __CYGWIN__)
 #include <sys/param.h>
+#else
+#include "windows-mingw/sys_param.h"
+#endif
+#if !(defined _WIN32 && ! defined __CYGWIN__)
 #include <sys/queue.h>
+#else
+#include "windows-mingw/sys_queue.h"
+#endif
 
diff --git a/libelftc/libelftc_bfdtarget.c b/libelftc/libelftc_bfdtarget.c
index 1b8dcbf5..29f561b7 100644
--- a/libelftc/libelftc_bfdtarget.c
+++ b/libelftc/libelftc_bfdtarget.c
@@ -26,3 +26,7 @@
 
+#if !(defined _WIN32 && ! defined __CYGWIN__)
 #include <sys/param.h>
+#else
+#include "windows-mingw/sys_param.h"
+#endif
 #include <libelf.h>
diff --git a/libelftc/make-toolchain-version b/libelftc/make-toolchain-version
index 1de0884f..40d18b31 100755
--- a/libelftc/make-toolchain-version
+++ b/libelftc/make-toolchain-version
@@ -10,3 +10,3 @@
 #
-buildhost=`uname -s`
+buildhost=`uname -s | cut -d_ -f1`
 elftcname="elftoolchain"