diff --git  libelf/_libelf.h libelf/_libelf.h
index 6d73d8f7..e6b6d543 100644
---  libelf/_libelf.h
+++ libelf/_libelf.h
@@ -31,3 +31,7 @@
 
+#if !(defined _WIN32 && ! defined __CYGWIN__)
 #include <sys/queue.h>
+#else
+#include "windows-mingw/sys_queue.h"
+#endif
 
diff --git  libelf/_libelf_ar.h libelf/_libelf_ar.h
index d1751a7c..e0be905a 100644
---  libelf/_libelf_ar.h
+++ libelf/_libelf_ar.h
@@ -36,3 +36,7 @@
 
+#if !(defined _WIN32 && ! defined __CYGWIN__)
 #include <ar.h>
+#else
+#include "windows-mingw/ar.h"
+#endif
 
diff --git  libelf/_libelf_config.h libelf/_libelf_config.h
index 2db8c369..bb8da0d4 100644
---  libelf/_libelf_config.h
+++ libelf/_libelf_config.h
@@ -165,2 +165,8 @@
 
+#if (defined _WIN32 && ! defined __CYGWIN__)
+#define LIBELF_ARCH      EM_386
+#define LIBELF_BYTEORDER ELFDATA2LSB
+#define LIBELF_CLASS     ELFCLASS32
+#endif
+
 /*
diff --git  libelf/elf_getident.c libelf/elf_getident.c
index 1a71b27c..6debf18b 100644
---  libelf/elf_getident.c
+++ libelf/elf_getident.c
@@ -26,3 +26,7 @@
 
+#if !(defined _WIN32 && ! defined __CYGWIN__)
 #include <ar.h>
+#else
+#include "windows-mingw/ar.h"
+#endif
 #include <assert.h>
diff --git  libelf/elf_next.c libelf/elf_next.c
index ad595a0a..2c3ea10c 100644
---  libelf/elf_next.c
+++ libelf/elf_next.c
@@ -26,3 +26,7 @@
 
+#if !(defined _WIN32 && ! defined __CYGWIN__)
 #include <ar.h>
+#else
+#include "windows-mingw/ar.h"
+#endif
 #include <assert.h>
diff --git  libelf/elf_phnum.c libelf/elf_phnum.c
index b08cc303..cbed1031 100644
---  libelf/elf_phnum.c
+++ libelf/elf_phnum.c
@@ -26,3 +26,7 @@
 
+#if !(defined _WIN32 && ! defined __CYGWIN__)
 #include <ar.h>
+#else
+#include "windows-mingw/ar.h"
+#endif
 #include <libelf.h>
diff --git  libelf/elf_rand.c libelf/elf_rand.c
index cd8774d3..d28cf11f 100644
---  libelf/elf_rand.c
+++ libelf/elf_rand.c
@@ -26,3 +26,7 @@
 
+#if !(defined _WIN32 && ! defined __CYGWIN__)
 #include <ar.h>
+#else
+#include "windows-mingw/ar.h"
+#endif
 #include <libelf.h>
diff --git  libelf/elf_scn.c libelf/elf_scn.c
index 4a067781..d4233777 100644
---  libelf/elf_scn.c
+++ libelf/elf_scn.c
@@ -26,3 +26,7 @@
 
+#if !(defined _WIN32 && ! defined __CYGWIN__)
 #include <sys/queue.h>
+#else
+#include "windows-mingw/sys_queue.h"
+#endif
 
diff --git  libelf/elf_shnum.c libelf/elf_shnum.c
index 66c134db..15709856 100644
---  libelf/elf_shnum.c
+++ libelf/elf_shnum.c
@@ -26,3 +26,7 @@
 
+#if !(defined _WIN32 && ! defined __CYGWIN__)
 #include <ar.h>
+#else
+#include "windows-mingw/ar.h"
+#endif
 #include <libelf.h>
diff --git  libelf/elf_shstrndx.c libelf/elf_shstrndx.c
index 3b2384cb..0b07fe78 100644
---  libelf/elf_shstrndx.c
+++ libelf/elf_shstrndx.c
@@ -26,3 +26,7 @@
 
+#if !(defined _WIN32 && ! defined __CYGWIN__)
 #include <ar.h>
+#else
+#include "windows-mingw/ar.h"
+#endif
 #include <libelf.h>
diff --git  libelf/elf_strptr.c libelf/elf_strptr.c
index c6a52661..4606b67e 100644
---  libelf/elf_strptr.c
+++ libelf/elf_strptr.c
@@ -26,3 +26,7 @@
 
+#if !(defined _WIN32 && ! defined __CYGWIN__)
 #include <sys/param.h>
+#else
+#include "windows-mingw/sys_param.h"
+#endif
 
diff --git  libelf/elf_update.c libelf/elf_update.c
index 6417afdc..64aacf8e 100644
---  libelf/elf_update.c
+++ libelf/elf_update.c
@@ -26,3 +26,7 @@
 
+#if !(defined _WIN32 && ! defined __CYGWIN__)
 #include <sys/param.h>
+#else
+#include "windows-mingw/sys_param.h"
+#endif
 #include <sys/stat.h>
diff --git  libelf/libelf.h libelf/libelf.h
index 8d7059b8..ffee3bef 100644
---  libelf/libelf.h
+++ libelf/libelf.h
@@ -33,2 +33,7 @@
 
+#if (defined _WIN32 && ! defined __CYGWIN__)
+typedef int gid_t;
+typedef int uid_t;
+#endif
+
 #include <elfdefinitions.h>
diff --git  libelf/libelf_memory.c libelf/libelf_memory.c
index cc393d16..172a10ae 100644
---  libelf/libelf_memory.c
+++ libelf/libelf_memory.c
@@ -26,3 +26,7 @@
 
+#if !(defined _WIN32 && ! defined __CYGWIN__)
 #include <ar.h>
+#else
+#include "windows-mingw/ar.h"
+#endif
 #include <assert.h>
diff --git  libelf/libelf_open.c libelf/libelf_open.c
index ea159c4d..e4c7c7b9 100644
---  libelf/libelf_open.c
+++ libelf/libelf_open.c
@@ -145,4 +145,8 @@ _libelf_open_object(int fd, Elf_Cmd c, int reporterror)
 	 */
+#if defined(S_ISSOCK)
 	if (!S_ISREG(mode) && !S_ISCHR(mode) && !S_ISFIFO(mode) &&
 	    !S_ISSOCK(mode)) {
+#else
+    if (!S_ISREG(mode) && !S_ISCHR(mode) && !S_ISFIFO(mode)) {
+#endif
 		LIBELF_SET_ERROR(ARGUMENT, 0);