diff --git a/libdw/dwarf_getsrclines.c b/libdw/dwarf_getsrclines.c
index 69e10c7b..473255dd 100644
--- a/libdw/dwarf_getsrclines.c
+++ b/libdw/dwarf_getsrclines.c
@@ -171,8 +171,8 @@ read_srclines (Dwarf *dbg,
      the stack.  Stack allocate some entries, only dynamically malloc
      when more than MAX.  */
 #define MAX_STACK_ALLOC 4096
-#define MAX_STACK_LINES (MAX_STACK_ALLOC / 2)
-#define MAX_STACK_FILES (MAX_STACK_ALLOC / 4)
+#define MAX_STACK_LINES (MAX_STACK_ALLOC / 32)
+#define MAX_STACK_FILES (MAX_STACK_ALLOC / 32)
 #define MAX_STACK_DIRS  (MAX_STACK_ALLOC / 16)
 
   /* Initial statement program state (except for stmt_list, see below).  */
@@ -329,7 +329,7 @@ read_srclines (Dwarf *dbg,
       const unsigned char *dirp = linep;
       while (dirp < lineendp && *dirp != 0)
 	{
-	  uint8_t *endp = memchr (dirp, '\0', lineendp - dirp);
+	  const uint8_t *endp = memchr (dirp, '\0', lineendp - dirp);
 	  if (endp == NULL)
 	    goto invalid_data;
 	  ++ndirs;
@@ -405,7 +405,7 @@ read_srclines (Dwarf *dbg,
       for (unsigned int n = 1; n < ndirlist; n++)
 	{
 	  dirarray[n].dir = (char *) linep;
-	  uint8_t *endp = memchr (linep, '\0', lineendp - linep);
+	  const uint8_t *endp = memchr (linep, '\0', lineendp - linep);
 	  assert (endp != NULL); // Checked above when calculating ndirlist.
 	  dirarray[n].len = endp - linep;
 	  linep = endp + 1;
@@ -793,7 +793,7 @@ read_srclines (Dwarf *dbg,
 	    case DW_LNE_define_file:
 	      {
 		char *fname = (char *) linep;
-		uint8_t *endp = memchr (linep, '\0', lineendp - linep);
+		const uint8_t *endp = memchr (linep, '\0', lineendp - linep);
 		if (endp == NULL)
 		  goto invalid_data;
 		size_t fnamelen = endp - linep;
